network/limitlistener/rejectingLimitListener.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2013 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// RejectingLimitListener is a modification of LimitListener in
+// "golang.org/x/net/netutil". The difference is that when the number of connections
+// exceeds the limit, RejectingLimitListener will accept and immediately close all
+// new connections.
+
+package limitlistener
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+)
+
+// RejectingLimitListener returns a Listener that accepts at most n simultaneous
+// connections from the provided Listener.
+func RejectingLimitListener(l net.Listener, n uint64) net.Listener {
+	return &rejectingLimitListener{
+		Listener: l,
+		sem:      make(chan struct{}, n),
+		done:     make(chan struct{}),
+	}
+}
+
+type rejectingLimitListener struct {
+	net.Listener
+	sem       chan struct{}
+	closeOnce sync.Once     // ensures the done chan is only closed once
+	done      chan struct{} // no values sent; closed when Close is called
+}
+
+func (l *rejectingLimitListener) release() {
+	<-l.sem
+}
+
+func (l *rejectingLimitListener) Accept() (net.Conn, error) {
+	for {
+		select {
+		case <-l.done:
+			return nil, errors.New("Accept() limit listener is closed")
```

Comment:
Why do you need the for loop, couldn't you return this error after calling `c.Close()`?

Reply:
We need the for loop to close all new connections if we are above the threshold.

Reply:
please elaborate. One `Accept` per upcoming connection, if `l.sem` is full then default branch closes it immediately, and there is no other connections until next iteration in the loop for the same client. What is the point?

Reply:
I think @algorandskiy was suggesting the remaining connections would be flushed by the server implicitly because it constantly checks for new connections. Therefore the loop to terminate them here is not required.